docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+with stale generation numbers: this would reduce any possible extra load on the safekeeper from stale pageservers,
+and provide feedback to stale pageservers that they should shut down.
+
+This is not required for safety: reads from a stale pageserver are functionally harmless and only
+waste system resources. Logical writes (i.e. updates to remote_persistent_lsn that can cause trimming)
+are handled on the pageserver side (see "WAL trim changes" above).
+
+Note that the safekeeper should only reject reads for stale _pageserver_ generations. Stale _attachment_
+generations are valid for reads, as a tenant may be multi-attached during a migration, where two different
+pageservers are both replaying the WALs for the same tenant.
+
+## Operational impact
+
+### Availability
+
+Coordination of generation numbers via the control plane introduce a dependency for certain
+operations:
+
+1. Starting new pageservers (or activating pageservers after a restart)
+2. Executing enqueued deletions
+3. Advertising updated `remote_consistent_lsn` to enable WAL trimming
+
+Item 1. would mean that some in-place restarts that previously would have resumed service even if the control plane were
+unavailable, will now not resume service to users until the control plane is available. We could
+avoid this by having a timeout on communication with the control plane, and after some timeout,
+resume service with the node's previous generation (assuming this was persisted to disk). However,
+this is unlikely to be needed as the control plane is already an essential & highly available component.
+
+Item 2. is a non-issue operationally: it's harmless to delay deletions, the only impact of objects pending deletion is
+the S3 capacity cost.
+
+Item 3. is an issue if safekeepers are low on disk space and the control plane is unavailable for a long time.
+
+For a managed service, the general approach should be to make sure we are monitoring & respond fast enough
+that control plane outages are bounded in time. The separate of console and control plane will also help
+to keep the control plane itself simple and robust.
+
+We should also implement an "escape hatch" config for node generation numbers, where in a major disaster outage,
+we may manually run pageservers with a hand-selected generation number, so that we can bring them online
+independently of a control plane.
+
+### Rollout
+
+Although there is coupling between components, we may deploy most of the new data plane components
+independently of the control plane: initially they can just use a static generation number.
+
+#### Phase 1
+
+The pageserver is deployed with some special config to:
+
+- Always act like everything is generation 1 and do not wait for a control plane issued generation on startup.
+- Skip the places in deletion and remote_consistent_lsn updates where we would call into control plane
+
+The storage broker will tolerate the timeline state omitting generation numbers.
```

Comment:
Storage broker has only been casually mentioned up to this point.

Reply:
Added a dedicated section in optimizations for the storage broker, and referenced it from here and a couple other places 6d59f5c1